ethexe/contracts/src/Router.sol

```diff
@@ -379,34 +377,51 @@ contract Router is IRouter, OwnableUpgradeable, ReentrancyGuardTransient {
 
     /* Helper private functions */
 
-    function _createProgram(bytes32 _codeId, bytes32 _salt) private returns (address) {
-        Storage storage router = _router();
+    function _createProgram(bytes32 _codeId, Storage storage router) private returns (address program) {
         require(router.genesisBlock.hash != bytes32(0), "router genesis is zero; call `lookupGenesisHash()` first");
 
         require(
             router.protocolData.codes[_codeId] == Gear.CodeState.Validated,
             "code must be validated before program creation"
         );
 
-        // Check for duplicate isn't necessary, because `Clones.cloneDeterministic`
-        // reverts execution in case of address is already taken.
-        address actorId =
-            Clones.cloneDeterministic(router.implAddresses.mirrorProxy, keccak256(abi.encodePacked(_codeId, _salt)));
+        address _mirrorAbi = router.implAddresses.mirrorAbi;
 
-        router.protocolData.programs[actorId] = _codeId;
-        router.protocolData.programsCount++;
+        assembly ("memory-safe") {
+            let contractSize := extcodesize(_mirrorAbi)
 
-        emit ProgramCreated(actorId, _codeId);
+            let ctr := mload(0x40)
+            mstore(ctr, shl(0xf0, 0x3d61))
+            let offset := 0x02
+            mstore(add(ctr, offset), shl(0xf0, contractSize))
+            offset := add(offset, 0x02)
+            mstore(add(ctr, offset), shl(0xc8, 0x80600b3d3981f3))
+            offset := add(offset, 0x07)
+            mstore(0x40, add(ctr, offset))
```

Comment:
it's still ineffective, we could do it just like in Clones: 
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/ca7a4e39de0860bbaadf95824207886e6de9fa64/contracts/proxy/Clones.sol#L82

Reply:
`(uint256(0x3d61000080600b3d3981f3) << uint256(168)) | (uint256(0xEEEE & 0xFFFF) << uint256(224))`

Reply:
`mstore(0x40, add(ctr, offset))` leads to unaligned free memory pointer